crates/autopilot/src/database/events.rs

```diff
@@ -41,7 +41,7 @@ pub fn contract_to_db_events(
 
 #[async_trait::async_trait]
 impl EventStoring<ContractEvent> for Postgres {
-    async fn last_event_block(&self) -> Result<BlockNumberHash> {
+    async fn last_event_blocks(&self) -> Result<Vec<BlockNumberHash>> {
```

Comment:
Instead of changing this interface can we:
- Make the initial update always be `MAX_REORG_DEPTH`? This way, we don't need to retrieve the block number and hash from `EventStoring`. It will be a little slower, but only need to be updated once at the beginning. My concern is that fetching a list of block may add a lot of complexity in some places (for example, with the Subgraph initializers - do we need to fetch a list of blocks and hashes as well?)
- Have the `EventUpdate` **internally** keep track of recent block number and hashes?

This way, we don't need to "leak" the reorg depth detection logic that is needed in the `EventUpdater` into the various implementations?

Reply:
- Make the initial update always be MAX_REORG_DEPTH?

My understanding was that in case of DB used as a Storage, we want to continue with event update from the block where we left of before the restart, so we have a persistent chain of blocks with events without gaps in DB.

- Have the EventUpdate internally keep track of recent block number and hashes?

Do you mean this: https://github.com/cowprotocol/services/pull/490/files#diff-60652ce9b004202bff1704ee459d23bc170eadb86f66cf5c0161a431149086c2R33
We have the internal track in EventHandler and it works fine for initialization of UniV3 and Balancer (because they are always empty on restart) but not for the `Postgres` because of the first point. I guess that's why we need to ask the storage about the latest handled blocks.

- do we need to fetch a list of blocks and hashes as well

Yes, I added that here: https://github.com/cowprotocol/services/pull/485/files#diff-4385cf81679c4e8ed7fdc0e5d73dfbd28e733bcb384bea9517fcdeb2bc39f705R61
Seems like not a big deal since it's done once on init.

Reply:
> My understanding was that in case of DB used as a Storage, we want to continue with event update from the block where we left of before the restart, so we have a persistent chain of blocks with events without block gaps in DB.

Right, my point is that, for the initial fetch, we can start on the `last_updated_block - MAX_REORG_DEPTH` instead of trying to find if there were any reorgs. Since the "initial fetch" would only happen on startup - its OK if it takes a bit longer.

My concern is that we don't necessarily have a list of `(block, hash)` for our events (since there are definitely blocks where none of the events we observe were emitted - meaning we can't necessarily figure out where the last reorg happened.

> EDIT: on a second thought, did you mean we can ask the Postgres about the last handled block only and then internally in EventHandler get the 25 blocks before that using our new function get_current_blocks, that would have to change the signature to something like:
async fn current_blocks(&self, last_block: u64) -> Result<Vec<BlockNumberHash>>;
?

Exactly!

Reply:
> Yes, I added that here: https://github.com/cowprotocol/services/pull/485/files#diff-4385cf81679c4e8ed7fdc0e5d73dfbd28e733bcb384bea9517fcdeb2bc39f705R61
Seems like not a big deal since it's done once on init.

Yes, this makes sense. I thought it would need to get a list of `(number, hash)` and not just one.

Reply:
I think this could work. Thanks. This means I can also drop the changes for `Postgres` introduced in https://github.com/cowprotocol/services/pull/485